File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Route.java
package com.bombbird.terminalcontrol.entities.sidstar;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.aircrafts.Aircraft;
import com.bombbird.terminalcontrol.entities.aircrafts.Departure;
import com.bombbird.terminalcontrol.entities.procedures.holding.HoldProcedure;
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint;
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen;
import com.bombbird.terminalcontrol.utilities.math.MathTools;
import org.json.JSONArray;
import org.json.JSONObject;

public class Route {
    private final RadarScreen radarScreen;

    private final Array<Waypoint> wpts;
    private final Array<int[]> restrictions;
    private final Array<Boolean> flyOver;
    private HoldProcedure holdProcedure;

    private String name;

    /** Basic constructor for Route, should only be called by other Route constructors */
    private Route() {
        radarScreen = TerminalControl.radarScreen;

        wpts = new Array<>();
        restrictions = new Array<>();
        flyOver = new Array<>();
    }

    /** Create a new Route based on STAR, for compatibility with older versions */
    public Route(Star star) {
        this();
        Array<String> inbound = star.getRandomInbound();
        if (!"HDG".equals(inbound.get(0).split(" ")[0])) {
            String[] data = inbound.get(0).split(" ");
            wpts.add(radarScreen.waypoints.get(data[1]));
            restrictions.add(new int[] {Integer.parseInt(data[2]), Integer.parseInt(data[3]), Integer.parseInt(data[4])});
            flyOver.add(data.length > 5 && data[5].equals("FO"));
        }

        wpts.addAll(star.getWaypoints());
        restrictions.addAll(star.getRestrictions());
        flyOver.addAll(star.getFlyOver());

        holdProcedure = new HoldProcedure(star);

        name = star.getName();
    }

    /** Create a new Route based on SID, for compatibility with older versions */
    public Route(Sid sid, String runway) {
        this();
        wpts.addAll(sid.getInitWpts(runway));
        restrictions.addAll(sid.getInitRestrictions(runway));
        flyOver.addAll(sid.getInitFlyOver(runway));
        wpts.addAll(sid.getWaypoints());
        restrictions.addAll(sid.getRestrictions());
        flyOver.addAll(sid.getFlyOver());
        Array<String> transition = sid.getRandomTransition();
        for (int i = 0; i < transition.size; i++) {
            String[] data = transition.get(i).split(" ");
            if (data[0].equals("WPT")) {
                //Waypoint
                wpts.add(TerminalControl.radarScreen.waypoints.get(data[1]));
                restrictions.add(new int[] {Integer.parseInt(data[2]), Integer.parseInt(data[3]), Integer.parseInt(data[4])});
                flyOver.add(data.length > 5 && data[5].equals("FO"));
            }
        }

        holdProcedure = new HoldProcedure();

        name = sid.getName();
    }

    /** Create new Route based on newly assigned STAR */
    public Route(Aircraft aircraft, Star star) {
        this();
        Array<String> inbound = star.getRandomInbound();
        for (int i = 0; i < inbound.size; i++) {
            if ("HDG".equals(inbound.get(i).split(" ")[0])) {
                aircraft.setHeading(Integer.parseInt(inbound.get(i).split(" ")[1]));
            } else {
                String[] data = inbound.get(i).split(" ");
                wpts.add(radarScreen.waypoints.get(data[1]));
                restrictions.add(new int[] {Integer.parseInt(data[2]), Integer.parseInt(data[3]), Integer.parseInt(data[4])});
                flyOver.add(data.length > 5 && data[5].equals("FO"));
            }
        }

        wpts.addAll(star.getWaypoints());
        restrictions.addAll(star.getRestrictions());
        flyOver.addAll(star.getFlyOver());

        String runway = null;
        for (int i = 0; i < star.getRunways().size; i++) {
            String rwy = star.getRunways().get(i);
            if (aircraft.getAirport().getLandingRunways().containsKey(rwy)) {
                runway = rwy;
                break;
            }
        }

        if (runway == null) throw new RuntimeException("Runway selected is null");

        wpts.addAll(star.getRwyWpts(runway));
        restrictions.addAll(star.getRwyRestrictions(runway));
        flyOver.addAll(star.getRwyFlyOver(runway));

        holdProcedure = new HoldProcedure(star);

        name = star.getName();
    }

    /** Create new Route based on newly assigned SID */
    public Route(Aircraft aircraft, Sid sid, String runway) {
        this();
        wpts.addAll(sid.getInitWpts(runway));
        restrictions.addAll(sid.getInitRestrictions(runway));
        flyOver.addAll(sid.getInitFlyOver(runway));
        wpts.addAll(sid.getWaypoints());
        restrictions.addAll(sid.getRestrictions());
        flyOver.addAll(sid.getFlyOver());
        Array<String> transition = sid.getRandomTransition();
        for (int i = 0; i < transition.size; i++) {
            String[] data = transition.get(i).split(" ");
            if (data[0].equals("WPT")) {
                //Waypoint
                wpts.add(TerminalControl.radarScreen.waypoints.get(data[1]));
                restrictions.add(new int[] {Integer.parseInt(data[2]), Integer.parseInt(data[3]), Integer.parseInt(data[4])});
                flyOver.add(data.length > 5 && data[5].equals("FO"));
            } else {
                //Outbound heading
                ((Departure) aircraft).setOutboundHdg(Integer.parseInt(data[MathUtils.random(1, data.length - 1)]));
            }
        }

        holdProcedure = new HoldProcedure();

        name = sid.getName();
    }

    /** Create new Route based on saved route, called only by other constructors */
    private Route(JSONObject jo) {
        this();

        JSONArray waypoints = jo.getJSONArray("waypoints");
        JSONArray restr = jo.getJSONArray("restrictions");
        JSONArray fo = jo.getJSONArray("flyOver");
        for (int i = 0; i < waypoints.length(); i++) {
            wpts.add(radarScreen.waypoints.get(waypoints.getString(i)));
            String[] data = restr.getString(i).split(" ");
            restrictions.add(new int[] {Integer.parseInt(data[0]), Integer.parseInt(data[1]), Integer.parseInt(data[2])});
            flyOver.add(fo.getBoolean(i));
        }

        holdProcedure = new HoldProcedure();

        name = jo.isNull("name") ? "null" : jo.getString("name");
    }

    /** Create new Route based on saved route and SID name */
    public Route(JSONObject jo, Sid sid) {
        this(jo);

        if ("null".equals(name) && sid != null) name = sid.getName();
    }

    /** Create new Route based on saved route and STAR name */
    public Route(JSONObject jo, Star star) {
        this(jo);
        holdProcedure = new HoldProcedure(star);

        name = star.getName();
    }

    /** Draws the lines between aircraft, waypoints with shapeRenderer */
    public void joinLines(int start, int end, int outbound) {
        Waypoint prevPt = null;
        int index = start;
        while (index < end) {
            Waypoint waypoint = getWaypoint(index);
            if (prevPt != null) {
                radarScreen.shapeRenderer.line(prevPt.getPosX(), prevPt.getPosY(), waypoint.getPosX(), waypoint.getPosY());
            }
            prevPt = waypoint;
            index++;
        }
        if (prevPt != null) {
            drawOutbound(prevPt.getPosX(), prevPt.getPosY(), outbound);
        }
    }

    /** Draws the outbound track from waypoint (if latMode is After waypoint, fly heading) */
    private void drawOutbound(float previousX, float previousY, int outbound) {
        if (outbound != -1 && previousX <= 4500 && previousX >= 1260 && previousY <= 3240 && previousY >= 0) {
            float outboundTrack = outbound - radarScreen.magHdgDev;
            float[] point = MathTools.pointsAtBorder(new float[] {1260, 4500}, new float[] {0, 3240}, previousX, previousY, outboundTrack);
            radarScreen.shapeRenderer.line(previousX, previousY, point[0], point[1]);
        }
    }

    public Array<Waypoint> getWaypoints() {
        return wpts;
    }

    public Waypoint getWaypoint(int index) {
        if (index >= wpts.size) {
            return null;
        }
        return wpts.get(index);
    }

    /** Calculates distance between remaining points, excluding distance between aircraft and current waypoint */
    public float distBetRemainPts(int nextWptIndex) {
        int currentIndex = nextWptIndex;
        float dist = 0;
        while (currentIndex < getWaypoints().size - 1) {
            dist += distBetween(currentIndex, currentIndex + 1);
            currentIndex++;
        }
        return dist;
    }

    /** Calculates distance between 2 waypoints in the route based on their indices */
    public float distBetween(int pt1, int pt2) {
        Waypoint waypoint1 = getWaypoint(pt1);
        Waypoint waypoint2 = getWaypoint(pt2);
        return MathTools.pixelToNm(MathTools.distanceBetween(waypoint1.getPosX(), waypoint1.getPosY(), waypoint2.getPosX(), waypoint2.getPosY()));
    }

    /** Returns an array of waypoints from start to end index inclusive */
    public Array<Waypoint> getRemainingWaypoints(int start, int end) {
        //Returns array of waypoints starting from index
        Array<Waypoint> newRange = new Array<>(wpts);
        if (end >= start) {
            if (start > 0) {
                newRange.removeRange(0, start - 1);
            }
            int newEnd = end - start;
            if (newEnd < newRange.size - 1) {
                newRange.removeRange(newEnd + 1, newRange.size - 1);
            }
        }
        return newRange;
    }

    public int findWptIndex(String wptName) {
        return wpts.indexOf(radarScreen.waypoints.get(wptName), false);
    }

    public int getWptMinAlt(String wptName) {
        return restrictions.get(findWptIndex(wptName))[0];
    }

    public int getWptMinAlt(int index) {
        return restrictions.get(index)[0];
    }

    public int getWptMaxAlt(String wptName) {
        return restrictions.get(findWptIndex(wptName))[1];
    }

    public int getWptMaxAlt(int index) {
        return restrictions.get(index)[1];
    }

    public int getWptMaxSpd(String wptName) {
        return restrictions.get(findWptIndex(wptName))[2];
    }

    public int getWptMaxSpd(int index) {
        return restrictions.get(index)[2];
    }

    public boolean getWptFlyOver(String wptName) {
        return flyOver.get(findWptIndex(wptName));
    }

    public HoldProcedure getHoldProcedure() {
        return holdProcedure;
    }

    public Array<int[]> getRestrictions() {
        return restrictions;
    }

    public Array<Boolean> getFlyOver() {
        return flyOver;
    }

    public String getName() {
        return name;
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/sidstar/Route.kt
package com.bombbird.terminalcontrol.entities.sidstar

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.aircrafts.Aircraft
import com.bombbird.terminalcontrol.entities.aircrafts.Departure
import com.bombbird.terminalcontrol.entities.approaches.Approach
import com.bombbird.terminalcontrol.entities.procedures.holding.HoldProcedure
import com.bombbird.terminalcontrol.entities.runways.Runway
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import com.bombbird.terminalcontrol.entities.zones.SidStarZone
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen
import com.bombbird.terminalcontrol.utilities.math.MathTools
import com.bombbird.terminalcontrol.utilities.math.MathTools.distanceBetween
import com.bombbird.terminalcontrol.utilities.math.MathTools.pixelToNm
import com.bombbird.terminalcontrol.utilities.math.MathTools.pointsAtBorder
import org.json.JSONObject

class Route private constructor() {
    private val radarScreen: RadarScreen = TerminalControl.radarScreen!!
    private var isStar = false
    val routeData = RouteData()
    val routeDataDynamic = RouteData()
    lateinit var holdProcedure: HoldProcedure
        private set
    var heading: Int
        private set
    private lateinit var sidStarZone: SidStarZone
    lateinit var name: String
        private set
    val size
        get() = routeDataDynamic.size
    val waypoints
        get() = routeDataDynamic.waypoints
    var apchTrans = ""
    val removedPoints = RouteData()

    /** Initializes a default heading */
    init {
        heading = -1
    }

    /** Create new Route based on newly assigned STAR  */
    constructor(aircraft: Aircraft, star: Star, changeSTAR: Boolean) : this() {
        isStar = true
        val inbound = star.randomInbound
        for (i in 0 until inbound.size) {
            if ("HDG" == inbound[i].split(" ".toRegex()).toTypedArray()[0]) {
                if (!changeSTAR) aircraft.heading = inbound[i].split(" ".toRegex()).toTypedArray()[1].toInt().toDouble()
                heading = aircraft.heading.toInt() + 180
            } else {
                val data = inbound[i].split(" ".toRegex()).toTypedArray()
                routeData.add(
                    radarScreen.waypoints[data[1]]!!,
                    intArrayOf(data[2].toInt(), data[3].toInt(), data[4].toInt()),
                    data.size > 5 && data[5] == "FO"

                )
            }
        }
        routeData.addAll(star.routeData)
        var runway: String? = null
        for (i in 0 until star.runways.size) {
            val rwy = star.runways[i]
            if (aircraft.airport.landingRunways.containsKey(rwy)) {
                runway = rwy
                break
            }
        }
        if (runway == null) {
            runway = if (!star.runways.isEmpty) star.runways.first() else null
        }
        if (runway == null) throw RuntimeException("Runway selected is null")
        routeData.addAll(star.getRwyWpts(runway), star.getRwyRestrictions(runway), star.getRwyFlyOver(runway))
        recalculateRouteData()
        holdProcedure = HoldProcedure(star)
        name = star.name
        calculateStarZone()
    }

    /** Create new Route based on newly assigned SID  */
    constructor(aircraft: Aircraft, sid: Sid, runway: String, climbRate: Int) : this() {
        isStar = false
        routeData.addAll(sid.getInitWpts(runway)!!, sid.getInitRestrictions(runway)!!, sid.getInitFlyOver(runway)!!)
        routeData.addAll(sid.routeData)
        val transition = sid.randomTransition
        if (transition == null) {
            val waypoints = routeData.waypoints
            val calculatedTrack =  if (waypoints.size >= 2) {
                val wpt1 = waypoints[waypoints.size - 2]
                val wpt2 = waypoints[waypoints.size - 1]
                MathTools.getRequiredTrack(wpt1.posX, wpt1.posY, wpt2.posX, wpt2.posY).toInt()
            } else -1
            (aircraft as Departure).outboundHdg = calculatedTrack
            heading = calculatedTrack
        } else for (i in 0 until transition.size) {
            val data = transition[i].split(" ".toRegex()).toTypedArray()
            if (data[0] == "WPT") {
                //Waypoint
                    routeData.add(
                        radarScreen.waypoints[data[1]]!!,
                        intArrayOf(data[2].toInt(), data[3].toInt(), data[4].toInt()),
                        data.size > 5 && data[5] == "FO"
                    )
            } else {
                //Outbound heading
                (aircraft as Departure).outboundHdg = data[MathUtils.random(1, data.size - 1)].toInt()
                heading = aircraft.outboundHdg
            }
        }
        recalculateRouteData()
        holdProcedure = HoldProcedure()
        name = sid.name
        loadSidZone(aircraft.airport.runways[runway], sid, climbRate)
    }

    /** Create new Route based on saved route, called only by other constructors  */
    private constructor(jo: JSONObject) : this() {
        val waypoints = jo.getJSONArray("waypoints")
        val restr = jo.getJSONArray("restrictions")
        val fo = jo.getJSONArray("flyOver")
        for (i in 0 until waypoints.length()) {
            val data = restr.getString(i).split(" ".toRegex()).toTypedArray()
            routeData.add(
                radarScreen.waypoints[waypoints.getString(i)]!!,
                intArrayOf(data[0].toInt(), data[1].toInt(), data[2].toInt()),
                fo.getBoolean(i)
            )
        }

        val removedWpts = jo.optJSONArray("removedWpts")
        val removedRestr = jo.optJSONArray("removedRestr")
        val removedFo = jo.optJSONArray("removedFo")
        if (removedWpts != null && removedRestr != null && removedFo != null) for (i in 0 until removedWpts.length()) {
            val data = removedRestr.getString(i).split(" ".toRegex()).toTypedArray()
            removedPoints.add(
                radarScreen.waypoints[removedWpts.getString(i)]!!,
                intArrayOf(data[0].toInt(), data[1].toInt(), data[2].toInt()),
                removedFo.getBoolean(i)
            )
        }

        holdProcedure = HoldProcedure()
        heading = jo.optInt("heading", -1)
        name = jo.optString("name", "null")
        apchTrans = jo.optString("apchTrans", "")
    }

    /** Create new Route based on saved route and SID name  */
    constructor(jo: JSONObject, sid: Sid, runway: Runway, climbRate: Int) : this(jo) {
        isStar = false
        if ("null" == name) name = sid.name
        loadSidZone(runway, sid, climbRate)
        recalculateRouteData()
    }

    /** Create new Route based on saved route and STAR name  */
    constructor(jo: JSONObject, star: Star) : this(jo) {
        isStar = true
        holdProcedure = HoldProcedure(star)
        name = star.name
        calculateStarZone()
        recalculateRouteData()
    }

    /** Loads sidStarZone for STAR routes  */
    private fun calculateStarZone() {
        sidStarZone = SidStarZone(this, false)
        sidStarZone.calculatePolygons(0)
    }

    /** Loads sidStarZone for SID routes  */
    private fun loadSidZone(runway: Runway?, sid: Sid?, climbRate: Int) {
        sidStarZone = SidStarZone(this, true)
        sidStarZone.calculatePolygons(routeData.waypoints.size - 1)
        if (runway != null && sid != null && climbRate > -1) sidStarZone.calculateDepRwyPolygons(runway, sid, climbRate)
    }

    /** Draws the lines between aircraft, waypoints with shapeRenderer  */
    fun joinLines(start: Int, end: Int, outbound: Int) {
        var prevPt: Waypoint? = null
        var index = start
        while (index < end) {
            val waypoint = getWaypoint(index)
            if (prevPt != null && waypoint != null) {
                radarScreen.shapeRenderer.line(prevPt.posX.toFloat(), prevPt.posY.toFloat(), waypoint.posX.toFloat(), waypoint.posY.toFloat())
            }
            prevPt = waypoint
            index++
        }
        if (prevPt != null) {
            drawOutbound(prevPt.posX.toFloat(), prevPt.posY.toFloat(), outbound)
        }
    }

    /** Draws the outbound track from waypoint (if latMode is After waypoint, fly heading)  */
    private fun drawOutbound(previousX: Float, previousY: Float, outbound: Int) {
        if (outbound != -1 && previousX <= 4500 && previousX >= 1260 && previousY <= 3240 && previousY >= 0) {
            val outboundTrack = outbound - radarScreen.magHdgDev
            val point = pointsAtBorder(floatArrayOf(1260f, 4500f), floatArrayOf(0f, 3240f), previousX, previousY, outboundTrack)
            radarScreen.shapeRenderer.line(previousX, previousY, point[0], point[1])
        }
    }

    /** Draws the sidStarZone boundaries  */
    fun drawPolygons() {
        for (polygon in sidStarZone.polygons) {
            radarScreen.shapeRenderer.polygon(polygon.transformedVertices)
        }
    }

    /** Checks whether supplied coordinates is within the sidStarZone of the route  */
    fun inSidStarZone(x: Float, y: Float, alt: Float): Boolean {
        return sidStarZone.contains(x, y, alt)
    }

    fun getWaypoint(index: Int): Waypoint? {
        return if (index >= routeDataDynamic.waypoints.size) {
            null
        } else routeDataDynamic.waypoints[index]
    }

    /** Calculates distance between remaining points, excluding distance between aircraft and current waypoint  */
    fun distBetRemainPts(nextWptIndex: Int): Float {
        var currentIndex = nextWptIndex
        var dist = 0f
        while (currentIndex < routeData.waypoints.size - 1) {
            if (getWaypoint(currentIndex + 1)?.isInsideRadar != true) break
            dist += distBetween(currentIndex, currentIndex + 1)
            currentIndex++
        }
        return dist
    }

    /** Calculates distance between 2 waypoints in the route based on their indices  */
    fun distBetween(pt1: Int, pt2: Int): Float {
        val waypoint1 = getWaypoint(pt1)
        val waypoint2 = getWaypoint(pt2)
        return if (waypoint1 != null && waypoint2 != null) pixelToNm(distanceBetween(waypoint1.posX.toFloat(), waypoint1.posY.toFloat(), waypoint2.posX.toFloat(), waypoint2.posY.toFloat())) else 0f
    }

    /** Returns an array of waypoints from start to end index inclusive  */
    fun getRemainingWaypoints(start: Int, end: Int): Array<Waypoint> {
        //Returns array of waypoints from index start to end
        val newRange = Array(routeDataDynamic.waypoints)
        if (end >= start) {
            if (start > 0) {
                newRange.removeRange(0, start - 1)
            }
            val newEnd = end - start
            if (newEnd < newRange.size - 1) {
                newRange.removeRange(newEnd + 1, newRange.size - 1)
            }
        }
        return newRange
    }

    fun findWptIndex(wptName: String?): Int {
        return routeDataDynamic.waypoints.indexOf(radarScreen.waypoints[wptName], false)
    }

    fun getWptMinAlt(wptName: String?): Int {
        if (wptName == null) return -1
        return routeDataDynamic.restrictions[findWptIndex(wptName)][0]
    }

    fun getWptMinAlt(index: Int): Int {
        return routeDataDynamic.restrictions[index][0]
    }

    fun getWptMaxAlt(wptName: String?): Int {
        if (wptName == null) return -1
        return routeDataDynamic.restrictions[findWptIndex(wptName)][1]
    }

    fun getWptMaxAlt(index: Int): Int {
        return routeDataDynamic.restrictions[index][1]
    }

    fun getWptMaxSpd(wptName: String?): Int {
        if (wptName == null) return -1
        return routeDataDynamic.restrictions[findWptIndex(wptName)][2]
    }

    fun getWptMaxSpd(index: Int): Int {
        return routeDataDynamic.restrictions[index][2]
    }

    fun getWptFlyOver(wptName: String?): Boolean {
        if (wptName == null) return false
        return routeDataDynamic.flyOver[findWptIndex(wptName)]
    }

    /** Removes the waypoints from currently cleared approach, returns a boolean whether the aircraft is currently flying to one of the removed points */
    fun removeApchWpts(apch: Approach, direct: Waypoint?): Boolean {
        if (apch.routeDataMap.isEmpty() || routeData.waypoints.isEmpty) return false //No points to remove
        val wpts = apch.routeDataMap[apchTrans]?.waypoints ?: return false //If approach transition is not found, return
        val firstIndex = routeData.waypoints.lastIndexOf(wpts.first(), false)
        if (firstIndex == -1) return false //Route does not contain points
        routeData.removeRange(firstIndex, size - 1)
        routeData.addAll(removedPoints)
        removedPoints.clear()
        apchTrans = ""
        sidStarZone.calculatePolygons(0)
        recalculateRouteData()
        return wpts.contains(direct, false)
    }

    /** Adds the appropriate transition waypoints for the selected approach, returns a boolean whether any points were added */
    fun addApchWpts(apch: Approach, direct: Waypoint?): Boolean {
        if (apch.routeDataMap.isEmpty()) return false //No points to add
        removedPoints.clear()
        val trip = apch.getNextPossibleTransition(direct, this)
        trip.second.waypoints.let {
            if (it.isEmpty) return@let
            val firstIndex = findWptIndex(it.first().name)
            if (firstIndex > -1) routeData.removeRange(firstIndex, size - 1)
            removedPoints.addAll(trip.second)
        }
        trip.first.waypoints.let {
            if (it.isEmpty) return false
            if (waypoints.contains(it.first(), false)) return false //If waypoint already inside, don't add
            apchTrans = trip.third?.name ?: "vector"
            routeData.addAll(trip.first)
        }
        sidStarZone.calculatePolygons(0)
        recalculateRouteData()
        return true
    }

    /** Performs a recalculation of the route waypoint restrictions, due to a change in waypoints */
    private fun recalculateRouteData() {
        routeDataDynamic.clear()
        var minAlt = -1
        var maxAlt = -1
        var maxSpd = -1
        val wptData = routeData.waypoints
        val restrData = routeData.restrictions
        val foData = routeData.flyOver
        for (i in 0 until routeData.size) {
            if (isStar) {
                //Loop through for maximum altitude, waypoints, flyover, max speed
                if ((restrData[i][1] != -1 && maxAlt > restrData[i][1]) || (maxAlt == -1 && restrData[i][1] != -1)) maxAlt = restrData[i][1]
                if ((restrData[i][2] != -1 && maxSpd > restrData[i][2]) || (maxSpd == -1 && restrData[i][2] != -1)) maxSpd = restrData[i][2]
                routeDataDynamic.add(wptData[i], intArrayOf(-1, maxAlt, maxSpd), foData[i])
            } else {
                //Loop through for minimum altitude, waypoints, flyover
                if ((restrData[i][0] != -1 && minAlt < restrData[i][0]) || (minAlt == -1 && restrData[i][0] != -1)) minAlt = restrData[i][0]
                routeDataDynamic.add(wptData[i], intArrayOf(minAlt, -1, -1), foData[i])
            }
        }
        for (i in routeData.size - 1 downTo 0) {
            if (isStar) {
                //Loop through for minimum altitude
                if ((restrData[i][0] != -1 && minAlt < restrData[i][0]) || (minAlt == -1 && restrData[i][0] != -1)) minAlt = restrData[i][0]
                routeDataDynamic.restrictions[i][0] = minAlt
            } else {
                //Loop through for maximum altitude, max speed
                if ((restrData[i][1] != -1 && maxAlt > restrData[i][1]) || (maxAlt == -1 && restrData[i][1] != -1)) maxAlt = restrData[i][1]
                if ((restrData[i][2] != -1 && maxSpd > restrData[i][2]) || (maxSpd == -1 && restrData[i][2] != -1)) maxSpd = restrData[i][2]
                routeDataDynamic.restrictions[i][1] = maxAlt
                routeDataDynamic.restrictions[i][2] = maxSpd
            }
        }
    }
}